arangosh/Restore/RestoreFeature.cpp

```diff
@@ -1306,65 +967,534 @@ arangodb::Result processInputDirectory(
+                          "&force=" + (options.force ? "true" : "false") +
+                          "&useEnvelope=" + (useEnvelope ? "true" : "false");
+  
+  std::unordered_map<std::string, std::string> headers;
+  headers.emplace(arangodb::StaticStrings::ContentTypeHeader,
+                  arangodb::StaticStrings::MimeTypeDump);
+
+  std::unique_ptr<SimpleHttpResult> response(
+      client.request(arangodb::rest::RequestType::PUT, url, buffer, bufferSize, headers));
+  arangodb::Result res = ::checkHttpResponse(client, response, "restoring data", "");
+  
+  if (res.fail()) {
+    // error
+    LOG_TOPIC("a595a", WARN, Logger::RESTORE)
+        << "Error while restoring data into collection '" << collectionName
+        << "': " << res.errorMessage();
+    // leave readOffset in place in readOffsets, because the job did not succeed
+    {
+      // store error
+      MUTEX_LOCKER(locker, sharedState->mutex);
+      sharedState->result = res;
     }
   } else {
-    // restore indexes first
-    res = ::restoreIndexes(httpClient, jobData);
-    if (res.ok() && jobData.options.importData) {
-      res = ::restoreData(httpClient, jobData);
+    // no error
+
+    {
+      MUTEX_LOCKER(locker, sharedState->mutex);
+      TRI_ASSERT(!sharedState->readOffsets.empty());
+
+#ifdef ARANGODB_ENABLE_FAILURE_TESTS
+      if (options.failOnUpdateContinueFile) {
+        auto it = sharedState->readOffsets.find(readOffset);
+        TRI_ASSERT(it != sharedState->readOffsets.end());
+        [[maybe_unused]] bool wasSynced = progressTracker.updateStatus(
+            collectionName, arangodb::RestoreFeature::CollectionStatus{arangodb::RestoreFeature::RESTORING, readOffset + (*it).second});
+        if (wasSynced) {
+          LOG_TOPIC("a87bf", WARN, Logger::RESTORE) << "triggered failure point at offset " << readOffset << "!";
+          FATAL_ERROR_EXIT_CODE(38); // exit with exit code 38 to report to the test frame work that this was an intentional crash
+        }
+      }
+#endif
+      sharedState->readOffsets.erase(readOffset);
     }
+
+    updateProgress();
+  }
+  
+  stats.totalBatches++;
+  stats.totalSent += bufferSize;
+  return res;
+}
+
+void RestoreFeature::RestoreJob::updateProgress() {
+  MUTEX_LOCKER(locker, sharedState->mutex);
+  
+  if (!sharedState->readOffsets.empty()) {
+    auto it = sharedState->readOffsets.begin();
+    size_t readOffset = (*it).first;
+
+    // progressTracker has its own lock
+    locker.unlock();
+    
+    progressTracker.updateStatus(
+          collectionName, arangodb::RestoreFeature::CollectionStatus{arangodb::RestoreFeature::RESTORING, readOffset});
+  } else if (sharedState->readCompleteInputfile) {
+    // we are done with restoring the entire collection
+    
+    // progressTracker has its own lock
+    locker.unlock();
+
+    progressTracker.updateStatus(
+        collectionName, arangodb::RestoreFeature::CollectionStatus{arangodb::RestoreFeature::RESTORED, 0});
```

Comment:
Why do we report 0 here?

Reply:
The actual value `0` does not matter here. The more important thing is that we store status `RESTORED`, which means "fully restored".
This is copy&paste from the old code, and should be just fine.

Reply:
Got it!